src/core/ext/transport/chttp2/transport/ping_promise.h

```diff
@@ -0,0 +1,149 @@
+    keepalive_timeout_ = keepalive_timeout;
+  }
+
+ private:
+  // Returns a promise that sleeps for the keepalive timeout and triggers the
+  // keepalive timeout based on whether some data is read within the keepalive
+  // timeout.
+  auto WaitForKeepAliveTimeout();
+
+  // Returns a promise that sends a keepalive ping and spawns the keepalive
+  // timeout promise. The promise resolves in the following scenarios:
+  // 1. Ping ack is received within the keepalive timeout (when the ping ack is
+  //     received, a call to GotData() is expected).
+  // 2. Ping ack is received after keepalive timeout (but before the ping
+  //     timeout). In this case, if there is some data received while
+  //     waiting for ping ack, the keepalive timeout is not triggered and the
+  //     promise resolves when SendPing() resolves.
+  // 3. No data is received within the keepalive timeout and the keepalive
+  //    timeout is triggered.
+  auto TimeoutAndSendPing();
+
+  // Returns a promise that determines if a keepalive ping needs to be sent and
+  // sends a keepalive ping if needed. The promise resolves when either a ping
+  // ack is received or the keepalive timeout is triggered.
+  auto MaybeSendKeepAlivePing();
+
+  auto WaitForData() {
+    return [this]() -> Poll<absl::Status> {
+      if (IsDataReceivedInLastCycle()) {
+        VLOG(2) << "WaitForData: Data received. Poll resolved";
+        return absl::OkStatus();
+      } else {
+        VLOG(2) << "WaitForData: Data not received. Poll pending";
+        waker_ = GetContext<Activity>()->MakeNonOwningWaker();
+        return Pending{};
+      }
+    };
+  }
+  // Needs to be called when data is read from the peer.
+  void ReceivedData() {
+    if (IsKeepAliveTimeoutTriggered()) {
+      VLOG(2)
+          << "KeepAlive timeout triggered. Not setting data_received_ to true";
+      return;
+    }
+    VLOG(2) << "Data received. Setting data_received_ to true";
+    state_ |= kDataReceivedInLastCycle;
+    auto waker = std::move(waker_);
+    waker.Wakeup();
+  }
+  auto SendPing() {
+    DCHECK(!IsDataReceivedInLastCycle());
+    return keep_alive_interface_->SendPing();
+  }
+
+  // If no data is received in the last keepalive_interval, we should send a
+  // keepalive ping. This also means that there can be scenarios where we would
+  // send one keepalive ping in ~(2*keepalive_interval).
+  bool NeedToSendKeepAlivePing() {
+    return (IsDataReceivedInLastCycle() == false) &&
+           (keep_alive_interface_->NeedToSendKeepAlivePing());
+  }
+
+  GPR_ATTRIBUTE_ALWAYS_INLINE_FUNCTION bool IsKeepAliveTimeoutTriggered()
+      const {
+    return (state_ & kKeepAliveTimeoutTriggered);
+  }
+  GPR_ATTRIBUTE_ALWAYS_INLINE_FUNCTION bool IsDataReceivedInLastCycle() const {
+    return (state_ & kDataReceivedInLastCycle);
+  }
+
+  std::unique_ptr<KeepAliveSystemInterface> keep_alive_interface_;
+  // bit to indicate if data is received in the last cycle
```

Comment:
since there's only two bits, and we have a bunch of >sizeof(uint8_t) things in this class, we can just use two bools for state and save all the bitops